### PR TITLE
core: add support to see if there are messages available for the curr…

### PR DESCRIPTION
**Pull request by DipSwitch**
Add support to check if there are messages queued for the current thread.

This will later be used for `-EWOULDBLOCK` error code.

### CODE REVIEW

**Comment by gebart** [comment]
Could you add a simple test case for the new function?

I suggest changing the name to `msg_avail` to match the naming of other similar functions in the repo (`cib_avail`, `tsrb_avail`)

**Comment by DipSwitch** [comment]
Addressed name change and unit test.

File: tests/thread_msg_avail/Makefile [hunk 1]
```diff
@@ -0,0 +1,8 @@
+APPLICATION = thread_msg_avail
+include ../Makefile.tests_common
+
+BOARD_INSUFFICIENT_MEMORY := stm32f0discovery
```
Comment (authmillenon) [inline tests/thread_msg_avail/Makefile]: Are you sure? This application just uses `core`. If `stm32f0discovery` does not fit `core` we really reevaluate support for this platform ;-)

Comment (DipSwitch) [inline tests/thread_msg_avail/Makefile]: I copied it from `thread_msg` but I guess it has insufficient memory there because of 3 threads with printf support, sec ;)

File: core/msg.c [hunk 2]
```diff
@@ -370,6 +370,25 @@ static int _msg_receive(msg_t *m, int block)
     DEBUG("This should have never been reached!\n");
 }
 
+int msg_avail(void)
+{
+    unsigned state = disableIRQ();
+    DEBUG("msg_available: %" PRIkernel_pid ": msg_available.\n",
+          sched_active_thread->pid);
+
+    tcb_t *me = (tcb_t*) sched_threads[sched_active_pid];
```
Comment (Kijewski) [inline core/msg.c]: Replace `sched_threads[sched_active_pid]` with `sched_active_thread`.

File: core/msg.c [hunk 3]
```diff
@@ -370,6 +370,25 @@ static int _msg_receive(msg_t *m, int block)
     DEBUG("This should have never been reached!\n");
 }
 
+int msg_avail(void)
+{
+    unsigned state = disableIRQ();
```
Comment (Kijewski) [inline core/msg.c]: No locking needed.

File: core/msg.c (hunk 2 above)
Comment (DipSwitch) [inline core/msg.c]: I copy/pasted the code from `_msg_receive` should it get changed there aswell?

Comment (Kijewski) [inline core/msg.c]: Yes, please, and in `msg_send_receive`, too. But that should go into a new commit. Could be that I am missing a use case where `sched_threads[sched_active_pid] != sched_active_thread`, but I think that these two need to match in all cases but for a short moment then switching threads.

Comment (gebart) [inline core/msg.c]: The `sched_threads[sched_active_pid]` changes should go in a separate PR. The reviewer needs to look up all write occurences of sched_active_thread and sched_active_pid to ensure that we are not introducing a bug (possibly a race while switching threads) in the message passing system.

Comment (DipSwitch) [inline core/msg.c]: @gebart So for this PR I keep using `sched_threads[sched_active_pid]` to be on the save side?

Comment (gebart) [inline core/msg.c]: I was primarily talking about the existing functions which use `sched_threads[sched_active_pid]`, since it depends on where the function is used if there will be any possible races or not. I guess @Kijewski might have an opinion here?

Comment (Kijewski) [inline core/msg.c]: In here you can use `sched_active_thread`. For the other functions we need to look if they [are/may be] used in interrupt handlers. I am quite sure they must not be called by an interrupt handler, because it would simply make no sense if I am not mistaken.

**Comment by OlegHahm** [comment]
@DipSwitch, can you address @Kijewski's latest small comment?

**Comment by DipSwitch** [comment]
Kijewski: Replace sched_threads[sched_active_pid] with sched_active_thread.

Me: I copy/pasted the code from _msg_receive should it get changed there aswell?

Kijewski: Yes, please, and in msg_send_receive, too. But that should go into a new commit. Could be that I am missing a use case where sched_threads[sched_active_pid] != sched_active_thread, but I think that these two need to match in all cases but for a short moment then switching threads.

Do I make a pull request for the other lines as well?

I'll squash if you validated the last commit addressing @Kijewski's points.

**Comment by OlegHahm** [comment]
Validated. Please squash.

**Comment by DipSwitch** [comment]
squashed

File: core/include/msg.h [hunk 4]
```diff
@@ -225,6 +225,13 @@ int msg_reply(msg_t *m, msg_t *reply);
 int msg_reply_int(msg_t *m, msg_t *reply);
 
 /**
+ * @brief Check how many messages are available in the message queue
+ *
+ * @return Number of messages available in our queue
```
Comment (Kijewski) [inline core/include/msg.h]:
```
@return -1 if the caller's message queue was not initialized.
```

**Comment by Kijewski** [comment]
Please add the line of documentation about the error case. You can squash immediately.

Other than that: ACK. Needs a seconds ACK, because it's a core change.

**Comment by DipSwitch** [comment]
Done

**Comment by DipSwitch** [comment]
Travis failed on `Ready for CI build` on msp430 and armv7

**Comment by DipSwitch** [comment]
Travis is happy

**Comment by OlegHahm** [comment]
ACK and go!
